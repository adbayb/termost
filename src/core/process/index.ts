import { spawn } from "child_process";

export const exec = async (command: string, options: ExecOptions = {}) => {
	const promise = new Promise<string>((resolve, reject) => {
		let stdout = "";
		let stderr = "";
		const [bin, ...args] = command.split(" ") as [string, ...string[]];

		const childProcess = spawn(bin, args, {
			cwd: options.cwd,
			shell: true,
			stdio: "pipe",
			env: {
				...process.env,
				// @note: make sure to force color display for spawned processes
				FORCE_COLOR: "1",
			},
		});

		childProcess.stdout.on("data", (chunk) => {
			stdout += chunk;
		});

		childProcess.stderr.on("data", (chunk) => {
			stderr += chunk;
		});

		childProcess.on("close", (exitCode) => {
			if (exitCode !== 0) {
				reject(stderr.trim());
			} else {
				resolve(stdout.trim());
			}
		});
	});

	try {
		return await promise;
	} catch (error) {
		throw new Error(error);
	}
};

type ExecOptions = {
	cwd?: string;
};
