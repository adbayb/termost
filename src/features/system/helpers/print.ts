import chalk from "chalk";

type FormatColor =
	| "red"
	| "green"
	| "yellow"
	| "blue"
	| "magenta"
	| "cyan"
	| "grey"
	| "black"
	| "white";

type FormatModifier =
	| "bold"
	| "italic"
	| "underline"
	| "strikethrough"
	| "uppercase"
	| "lowercase";

const chalkByFormatColor = {
	red: "redBright",
	green: "greenBright",
	yellow: "yellowBright",
	blue: "blueBright",
	magenta: "magentaBright",
	cyan: "cyanBright",
	grey: "blackBright",
	black: "black",
	white: "white",
} as const;

const chalkByModifier = {
	bold: "bold",
	italic: "italic",
	underline: "underline",
	strikethrough: "strikethrough",
} as const;

type FormatFunction<ReturnValue = string> = (
	message: string,
	options?: {
		color?: FormatColor;
		modifier?: FormatModifier | Array<FormatModifier>;
	}
) => ReturnValue;

export const format: FormatFunction = (message, { color, modifier } = {}) => {
	let transformer: chalk.Chalk = chalk;

	if (color) {
		transformer = transformer[chalkByFormatColor[color]];
	}

	if (modifier) {
		const processOneModifier = (mod: FormatModifier) => {
			if (mod === "uppercase") {
				message = message.toUpperCase();
			} else if (mod === "lowercase") {
				message = message.toLowerCase();
			} else {
				transformer = transformer[chalkByModifier[mod]];
			}
		};

		if (Array.isArray(modifier)) {
			for (const mod of modifier) {
				processOneModifier(mod);
			}
		} else {
			processOneModifier(modifier);
		}
	}

	return transformer(message);
};

export const print: FormatFunction<void> = (message, options) => {
	console.log(format(message, options));
};

type BannerType = "success" | "error" | "warning" | "information";

const bannerPropertiesByType: Record<
	BannerType,
	{ icon: string; color: FormatColor }
> = {
	error: {
		icon: "❌",
		color: "red",
	},
	information: {
		icon: "ℹ️ ",
		color: "blue",
	},
	success: {
		icon: "✅",
		color: "green",
	},
	warning: {
		icon: "⚠️ ",
		color: "yellow",
	},
};

export const banner = (type: BannerType, message: string) => {
	const { icon, color } = bannerPropertiesByType[type];

	print(`${icon} ${message}`, { color, modifier: "bold" });
};
