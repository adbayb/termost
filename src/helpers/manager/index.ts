import { CommandName } from "../../features/types";
import { createQueue } from "./queue";

export const getManager = (commandName: CommandName): Manager => {
	const manager = managerCollection[commandName];

	if (!manager) {
		throw new Error(
			`No manager has been instantiated for the \`${commandName}\` command`
		);
	}

	return manager;
};

export const createManager = (commandName: CommandName): Manager => {
	const instructions = createQueue<Instruction>();

	const manager: Manager = {
		// addValue,
		// addOption,
		// getValue,
		// getOption,
		addInstruction(instruction) {
			instructions.enqueue(instruction);
		},
		async run() {
			while (!instructions.isEmpty()) {
				const task = instructions.dequeue();

				if (task) {
					await task();
				}
			}
		},
	};

	managerCollection[commandName] = manager;

	return manager;
};

type Instruction = () => Promise<void>;

type Manager = {
	addInstruction(instruction: Instruction): void;
	run(): Promise<void>;
};

const managerCollection: Record<CommandName, Manager> = {};
