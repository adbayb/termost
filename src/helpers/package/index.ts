import fs from "fs";
import path from "path";
import { PackageMetadata } from "../../types";

export const findNearestPackageJson = (pathname: string): PackageMetadata => {
	const packagePathname = path.resolve(pathname, "package.json");

	if (isFileExists(packagePathname)) {
		return require(packagePathname);
	}

	return findNearestPackageJson(path.resolve(pathname, ".."));
};

const isFileExists = (pathname: string) => {
	try {
		fs.accessSync(pathname, fs.constants.F_OK);

		return true;
	} catch {
		return false;
	}
};

export const getPackageMetadata = () => {
	const mainFilename = require.main?.filename;

	if (!mainFilename) {
		throw new Error(
			"Termost was unable to retrieve automatically the package name and version. To fix it, use `termost({ name, description, version })` to define them manually."
		);
	}

	return findNearestPackageJson(path.dirname(mainFilename));
};
