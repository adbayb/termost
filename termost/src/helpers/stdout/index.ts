import chalk from "chalk";

/**
 * A helper to format an arbitrary text as a message input
 * @param message The text to display
 * @param options The configuration object to control the formatting properties
 * @returns The formatted text
 */
export const format = (
	message: string,
	{
		color,
		modifier,
	}: {
		color?: Color;
		modifier?: Modifier | Modifier[];
	} = {},
): string => {
	let transformer: chalk.Chalk = chalk;

	if (color) {
		transformer = transformer[chalkByFormatColor[color]];
	}

	if (modifier) {
		const processOneModifier = (mod: Modifier) => {
			if (mod === "uppercase") {
				message = message.toUpperCase();
			} else if (mod === "lowercase") {
				message = message.toLowerCase();
			} else {
				transformer = transformer[chalkByModifier[mod]];
			}
		};

		if (Array.isArray(modifier)) {
			for (const mod of modifier) {
				processOneModifier(mod);
			}
		} else {
			processOneModifier(modifier);
		}
	}

	return transformer(message);
};

/**
 * An opinionated helper to display arbitrary text on the console
 * @param message The text to display. Use an array if you need to display a message in several lines
 * @param options The configuration object to define the display type and/or override the default label
 */
export const message = (
	content: string[] | string,
	{
		label,
		type = "information",
	}: { type?: MessageType; label?: string } = {},
) => {
	const { color, defaultLabel, icon, method } = formatPropertiesByType[type];
	const messages = typeof content === "string" ? [content] : content;

	method(
		format(`\n${icon} ${label ?? defaultLabel}`, {
			color,
			modifier: "bold",
		}),
	);

	for (const msg of messages) {
		method(format(`   ${msg}`, { color }));
	}
};

const formatPropertiesByType = {
	error: {
		color: "red",
		defaultLabel: "Error",
		icon: "❌",
		method: console.error,
	},
	information: {
		color: "blue",
		defaultLabel: "Information",
		icon: "ℹ️ ",
		method: console.info,
	},
	success: {
		color: "green",
		defaultLabel: "Success",
		icon: "✅",
		method: console.log,
	},
	warning: {
		color: "yellow",
		defaultLabel: "Warning",
		icon: "⚠️ ",
		method: console.warn,
	},
} as const;

const chalkByFormatColor = {
	red: "redBright",
	green: "greenBright",
	yellow: "yellowBright",
	blue: "blueBright",
	magenta: "magentaBright",
	cyan: "cyanBright",
	grey: "blackBright",
	black: "black",
	white: "white",
} as const;

const chalkByModifier = {
	bold: "bold",
	italic: "italic",
	underline: "underline",
	strikethrough: "strikethrough",
} as const;

type MessageType = "error" | "information" | "success" | "warning";

type Color =
	| "black"
	| "blue"
	| "cyan"
	| "green"
	| "grey"
	| "magenta"
	| "red"
	| "white"
	| "yellow";

type Modifier =
	| "bold"
	| "italic"
	| "lowercase"
	| "strikethrough"
	| "underline"
	| "uppercase";
