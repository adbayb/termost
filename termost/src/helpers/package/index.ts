import process from "node:process";
import { resolve } from "node:path";
import { createRequire } from "node:module";
import { accessSync, constants } from "node:fs";

import type { PackageMetadata } from "../../types";

const binCallerPath = process.argv[1] ?? "";
const require = createRequire(binCallerPath);

export const getPackageMetadata = (
	pathname: string = binCallerPath,
): PackageMetadata => {
	try {
		const packagePathname = resolve(pathname, "package.json");

		if (isFileExists(packagePathname)) {
			return require(packagePathname) as PackageMetadata;
		}

		return getPackageMetadata(resolve(pathname, ".."));
	} catch (error) {
		throw new Error(
			`Termost was unable to retrieve automatically the package name and version. To fix it, use \`termost({ name, description, version })\` to define them manually.\nMore details: ${String(error)}.`,
		);
	}
};

const isFileExists = (pathname: string) => {
	try {
		accessSync(pathname, constants.F_OK);

		return true;
	} catch {
		return false;
	}
};
