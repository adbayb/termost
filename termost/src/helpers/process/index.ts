import process from "node:process";
import { spawn } from "node:child_process";

export const exec = async (command: string, options: ExecOptions = {}) => {
	const { cwd, hasLiveOutput = false } = options;

	return new Promise<string>((resolve, reject) => {
		let stdout = "";
		let stderr = "";
		const [bin, ...arguments_] = command.split(" ") as [string, ...string[]];

		// eslint-disable-next-line sonarjs/os-command
		const childProcess = spawn(bin, arguments_, {
			cwd,
			env: {
				// eslint-disable-next-line n/no-process-env
				...process.env,
				// @note: make sure to force color display for spawned processes
				FORCE_COLOR: "1",
			},
			shell: true,
			stdio: hasLiveOutput ? "inherit" : "pipe",
		});

		childProcess.stdout?.on("data", (chunk: string) => {
			stdout += chunk;
		});

		childProcess.stderr?.on("data", (chunk: string) => {
			stderr += chunk;
		});

		childProcess.on("close", (exitCode) => {
			if (exitCode !== 0) {
				const output = `${stderr}${stdout}`;

				reject(new Error(output.trim()));
			} else {
				resolve(stdout.trim());
			}
		});
	});
};

type ExecOptions = {
	cwd?: string;
	hasLiveOutput?: boolean;
};
