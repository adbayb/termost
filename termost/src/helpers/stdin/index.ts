import process from "node:process";

export const getArguments = () => {
	const parameters = process.argv.slice(2);
	let command: string | undefined;
	const operands: string[] = [];
	const options: Record<string, boolean | number | string> = {};
	let currentOptionName: string | undefined;

	const addOptimisticOption = (name: string, value?: boolean | string) => {
		if (value) {
			options[name] = typeof value === "string" ? castValue(value) : true;
		} else {
			currentOptionName = name;
		}
	};

	const flushOptimisticOption = () => {
		if (currentOptionName) {
			options[currentOptionName] = true;
			currentOptionName = undefined;
		}
	};

	for (const parameter of parameters) {
		const shortFlagMatchResult = SHORT_FLAG_REGEX.exec(parameter)?.groups;
		const longFlagMatchResult = LONG_FLAG_REGEX.exec(parameter)?.groups;

		if (shortFlagMatchResult?.name) {
			flushOptimisticOption();

			const optionFlags = [...shortFlagMatchResult.name];
			const lastIndex = optionFlags.length - 1;

			optionFlags.forEach((flag, index) => {
				addOptimisticOption(
					flag,
					lastIndex === index ? undefined : true,
				);
			});
		} else if (longFlagMatchResult?.name) {
			flushOptimisticOption();
			addOptimisticOption(
				longFlagMatchResult.name,
				longFlagMatchResult.value,
			);
		} else if (currentOptionName) {
			options[currentOptionName] = castValue(parameter);
			currentOptionName = undefined;
		} else if (!command) {
			command = parameter;
		} else {
			operands.push(parameter);
		}
	}

	flushOptimisticOption();

	return { command, operands, options };
};

const SHORT_FLAG_REGEX = /^-(?<name>(?!-).*)$/;
const LONG_FLAG_REGEX = /^--(?<name>.*?)(?:=(?<value>.+))?$/;

const castValue = (value: string) => {
	try {
		return JSON.parse(value) as boolean | number | string;
	} catch {
		return value;
	}
};
