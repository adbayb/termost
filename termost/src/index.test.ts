import { describe, expect, test } from "vitest";

import { exec } from "./helpers/process";

describe("termost", () => {
	test("should display `version`", async () => {
		const longFlagOutput = await safeExec(
			"pnpm --filter @examples/default start --version",
		);

		const shortFlagOutput = await safeExec(
			"pnpm --filter @examples/default start -v",
		);

		expect(longFlagOutput).toMatchSnapshot();
		expect(shortFlagOutput).toMatchSnapshot();
	});

	test("should display `help`", async () => {
		const longFlagOutput = await safeExec(
			"pnpm --filter @examples/default start --help",
		);

		const shortFlagOutput = await safeExec(
			"pnpm --filter @examples/default start -h",
		);

		expect(longFlagOutput).toMatchSnapshot();
		expect(shortFlagOutput).toMatchSnapshot();
	});

	test("should display `help` given empty command", async () => {
		const rootCommand = await safeExec(
			"pnpm --filter @examples/empty start",
		);

		const buildCommand = await safeExec(
			"pnpm --filter @examples/empty start build",
		);

		const buildCommandWithOption = await safeExec(
			"pnpm --filter @examples/empty start build --option test",
		);

		const watchCommand = await safeExec(
			"pnpm --filter @examples/empty start watch",
		);

		expect(rootCommand).toMatchSnapshot();
		expect(buildCommand).toMatchSnapshot();
		expect(buildCommandWithOption).toMatchSnapshot();
		expect(watchCommand).toMatchSnapshot();
	});

	test("should handle `validation`", async () => {
		await expect(async () =>
			safeExec("pnpm --filter @examples/validation start:test -o error"),
		).rejects.toThrow(/Invalid option->input/);
		await expect(async () =>
			safeExec("pnpm --filter @examples/validation start:test"),
		).rejects.toThrow(/Invalid task->input/);
	});

	test("should handle `command` api", async () => {
		const helpOutput = await safeExec(
			"pnpm --filter @examples/command start --help",
		);

		const buildOutput = await safeExec(
			"pnpm --filter @examples/command start build",
		);

		const watchOutput = await safeExec(
			"pnpm --filter @examples/command start watch",
		);

		const buildSharedFlagOutput = await safeExec(
			"pnpm --filter @examples/command start build --global --local hello",
		);

		const watchSharedFlagOutput = await safeExec(
			"pnpm --filter @examples/command start watch --global",
		);

		const buildHelpOutput = await safeExec(
			"pnpm --filter @examples/command start build --help",
		);

		const watchHelpOutput = await safeExec(
			"pnpm --filter @examples/command start watch --help",
		);

		expect(helpOutput).toMatchSnapshot();
		expect(buildOutput).toMatchSnapshot();
		expect(watchOutput).toMatchSnapshot();
		expect(buildSharedFlagOutput).toMatchSnapshot();
		expect(watchSharedFlagOutput).toMatchSnapshot();
		expect(buildHelpOutput).toMatchSnapshot();
		expect(watchHelpOutput).toMatchSnapshot();
	});

	test("should handle `option` api", async () => {
		const output = await safeExec("pnpm --filter @examples/option start");

		expect(output).toMatchSnapshot();
	});

	test("should handle `task` api", async () => {
		const output = await safeExec("pnpm --filter @examples/task start");

		expect(output).toMatchSnapshot();
	});
});

/**
 * A test utility to strip contextual information (including absolute paths) output by `pnpm --filter run` command.
 * It allows to run tests whatever the testing environment (CI, local, ...).
 * @param command - The command to run.
 * @returns The generic command output.
 * @example
 * const shortFlagOutput = await safeExec(
 * 	"pnpm --filter @examples/default start -h",
 * );
 */
const safeExec = async (command: string) => {
	const output = await exec(command);

	return output.toString().split("\n").slice(3).join("\n");
};
