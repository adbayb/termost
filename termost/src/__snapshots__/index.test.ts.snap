// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`termost > should display \`help\` 1`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/default[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mProgram description placeholder. Program name and version are retrieved from your \`package.json\`. You can override this automatic retrieval by using the \`termost({ name, description, version })\` builder form.[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help        [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version     [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m-f, --flag        [32m[37m A super useful CLI flag[37m[39m"
`;

exports[`termost > should display \`help\` 2`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/default[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mProgram description placeholder. Program name and version are retrieved from your \`package.json\`. You can override this automatic retrieval by using the \`termost({ name, description, version })\` builder form.[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help        [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version     [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m-f, --flag        [32m[37m A super useful CLI flag[37m[39m"
`;

exports[`termost > should display \`help\` given empty command 1`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/empty[32m[37m <command> [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mExample to showcase empty \`command\` fallback[37m[39m
[33m[33m[1m[4m
COMMANDS:[24m[22m[33m[39m
[37m[37m  [32m[32mbuild           [32m[37m Transpile and bundle in production mode[37m[39m
[37m[37m  [32m[32mwatch           [32m[37m Rebuild your assets on any code change[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help      [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version   [32m[37m Print the version[37m[39m"
`;

exports[`termost > should display \`help\` given empty command 2`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/empty build[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mTranspile and bundle in production mode[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help      [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version   [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m--longOption    [32m[37m Useful CLI flag[37m[39m"
`;

exports[`termost > should display \`help\` given empty command 3`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/empty build[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mTranspile and bundle in production mode[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help      [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version   [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m--longOption    [32m[37m Useful CLI flag[37m[39m"
`;

exports[`termost > should display \`help\` given empty command 4`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/empty watch[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mRebuild your assets on any code change[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help      [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version   [32m[37m Print the version[37m[39m"
`;

exports[`termost > should display \`version\` 1`] = `"0.0.0"`;

exports[`termost > should display \`version\` 2`] = `"0.0.0"`;

exports[`termost > should handle \`command\` api 1`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/command[32m[37m <command> [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mExample to showcase the \`command\` API[37m[39m
[33m[33m[1m[4m
COMMANDS:[24m[22m[33m[39m
[37m[37m  [32m[32mbuild             [32m[37m Transpile and bundle in production mode[37m[39m
[37m[37m  [32m[32mwatch             [32m[37m Rebuild your assets on any code change[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help        [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version     [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m--global          [32m[37m Shared flag between commands[37m[39m"
`;

exports[`termost > should handle \`command\` api 2`] = `
"[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👋 Hello, I'm the build command[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Shared global flag = false[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Local command flag = local-value[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Context value = {"globalFlag":false,"localFlag":"local-value"}[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Argv value = {"command":"build","operands":[],"options":{}}[34m[39m"
`;

exports[`termost > should handle \`command\` api 3`] = `
"[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👋 Hello, I'm the watch command[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Shared global flag = false[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Context value = {"globalFlag":false}[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Argv value = {"command":"watch","operands":[],"options":{}}[34m[39m"
`;

exports[`termost > should handle \`command\` api 4`] = `
"[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👋 Hello, I'm the build command[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Shared global flag = true[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Local command flag = hello[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Context value = {"globalFlag":true,"localFlag":"hello"}[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Argv value = {"command":"build","operands":[],"options":{"global":true,"local":"hello"}}[34m[39m"
`;

exports[`termost > should handle \`command\` api 5`] = `
"[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👋 Hello, I'm the watch command[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Shared global flag = true[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Context value = {"globalFlag":true}[34m[39m
[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   👉 Argv value = {"command":"watch","operands":[],"options":{"global":true}}[34m[39m"
`;

exports[`termost > should handle \`command\` api 6`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/command build[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mTranspile and bundle in production mode[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help        [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version     [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m--global          [32m[37m Shared flag between commands[37m[39m
[37m[37m  [32m[32m--local           [32m[37m Local command flag[37m[39m"
`;

exports[`termost > should handle \`command\` api 7`] = `
"[33m[33m[1m[4m
USAGE:[24m[22m[33m[39m
[37m[37m[32m[32m@examples/command watch[32m[37m [...options][37m[39m
[33m[33m[1m[4m
DESCRIPTION:[24m[22m[33m[39m
[37m[37mRebuild your assets on any code change[37m[39m
[33m[33m[1m[4m
OPTIONS:[24m[22m[33m[39m
[37m[37m  [32m[32m-h, --help        [32m[37m Display the help center[37m[39m
[37m[37m  [32m[32m-v, --version     [32m[37m Print the version[37m[39m
[37m[37m  [32m[32m--global          [32m[37m Shared flag between commands[37m[39m"
`;

exports[`termost > should handle \`option\` api 1`] = `
"[34m[34m[1m
ℹ️  Information[22m[34m[39m
[34m[34m   {
  "optionWithAlias": 0,
  "optionWithoutAlias": "defaultValue"
}[34m[39m"
`;

exports[`termost > should handle \`task\` api 1`] = `
"[33m❯[39m Task with returned value (persisted)
[32m✔[39m Task with returned value (persisted)
[33m❯[39m Task with side-effect only (no persisted value)
[32m✔[39m Task with side-effect only (no persisted value)
[33m❯[39m Task can also access other persisted task values
[32m✔[39m Task can also access other persisted task values
[33m❯[39m Or even execute external commands thanks to its provided helpers
[32m✔[39m Or even execute external commands thanks to its provided helpers
[33m❯[39m A task can have a dynamic label generated from contextual values: small
[32m✔[39m A task can have a dynamic label generated from contextual values: small
[34m[34m[1m
ℹ️  Label & console output[22m[34m[39m
[34m[34m   If you don't specify a label, the handler is executed in "live mode" (the output is not hidden by the label and is displayed gradually).[34m[39m
[34m[34m[1m
ℹ️  Context management[22m[34m[39m
[34m[34m   A task with a specified "key" can be retrieved here. Size = 45. If no "key" was specified the task returned value cannot be persisted across program instructions.[34m[39m
[34m[34m[1m
ℹ️  Output formatting[22m[34m[39m
[34m[34m   The \`message\` helpers can be used to display task content in a nice way[34m[39m
[32m[32m[1m
✅ Success[22m[32m[39m
[32m[32m   The \`message\` helpers can be used to display task content in a nice way[32m[39m
[34m[34m[1m
ℹ️  👋 You can also customize the label[22m[34m[39m
[34m[34m   The \`message\` helpers can be used to display task content in a nice way[34m[39m
[34m[34m[1m
ℹ️  👋 You can also customize the label[22m[34m[39m
[34m[34m   I support also[34m[39m
[34m[34m   multilines[34m[39m
[34m[34m   with array input[34m[39m
[37m[37m[3m[9m[1m
You can also have a total control on the formatting through the \`format\` helper.[22m[29m[23m[37m[39m
{
  "size": 45,
  "computedFromOtherTaskValues": "small",
  "execOutput": "Hello from shell"
}"
`;
